feat(Timestamp): create component

thatblindgeye opened the pull request:
Closes #4801 

[Timestamp preview build](https://patternfly-pr-4996.surge.sh/components/timestamp)

Some notes/items to get feedback on:

- I noticed I still need to remove the CSS variable for textalign; depending on whether there are suggested changes I can lump that in with making those updates
- Originally I had used the `<time>` element for the timestamp text, but per MDN's [time element documentation](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/time), if the `datetime` attribute is not provided then the <time> element cannot have any child elements (and must be one of the valid date/time formats on the page). Based on the design docs it seems like the timestamp usage might not really require the time element, but it might be worth looking into offering as an opt-in down the line to take advantage of (per MDN, for "better search engine results or custom features such as reminders").
- Currently the timestamp has the pointer cursor on hover, but GitHub for example doesn't change the cursor when hovering over their timestamps. I figured a pointer might better convey that something will occur on hover (a tooltip/popover appearing), but would appreciate feedback on what others think.
- Because the timestamp would have a tooltip/popover to display the UTC (or some other custom, alternative datetime value), would it make sense to have this component behave more similarly to the [description list with term help text example](https://www.patternfly.org/v4/components/description-list#term-help-text) by making the timestamp a button? Tooltip content doesn't always get announced by assistive technologies on static elements, and per the current [tooltip ally doc](https://github.com/patternfly/patternfly-org/blob/main/packages/v4/patternfly-docs/content/accessibility/tooltip/tooltip.md) it's not recommended to place tooltips on static elements, but this may be something to decide on/update down the line.

patternfly-build commented:
Preview: https://patternfly-pr-4996.surge.sh

mattnolting commented on src/patternfly/components/Timestamp/timestamp.scss:
```diff
@@ -0,0 +1,24 @@
+  --pf-c-timestamp--Color: var(--pf-global--Color--200);
+  --pf-c-timestamp--TextAlign: center;
+  --pf-c-timestamp--Cursor: pointer;
+  --pf-c-timestamp--TextDecorationLine: underline;
+  --pf-c-timestamp--TextDecorationStyle: dashed;
+  --pf-c-timestamp--TextDecorationThickness: var(--pf-global--BorderWidth--sm);
+  --pf-c-timestamp--TextUnderlineOffset: #{pf-size-prem(4px)};
```
> Because the timestamp would have a tooltip/popover to display the UTC (or some other custom, alternative datetime value), would it make sense to have this component behave more similarly to the [description list with term help text example](https://www.patternfly.org/v4/components/description-list#term-help-text) by making the timestamp a button?

Looks to me like the design guidance presents these styles when the timestamp has a tooltip/popover. I'd move these to a `.pf-m-help-text` modifier.

mattnolting reviewed (commented):
Nice work @thatblindgeye! Left some feedback 👍

mcoker commented on src/patternfly/components/Timestamp/timestamp.hbs:
```diff
@@ -0,0 +1,6 @@
+<span class="pf-c-timestamp{{#if timestamp--modifier}} {{timestamp--modifier}}{{/if}}"
```
@mattnolting yeah @blfetche and I chatted about that and how we take our best stab at the intended structure and how we weigh potentially bloated markup versus being limited in updating the markup later as we have design changes that may introduce a breaking change. We looked around at other design system designs and saw some icons and things next to the timestamp, so figured it's fair we might put stuff beside it. That said, I don't feel strongly about it one way or another - hopefully we'll iron out the future design requirements while it's beta. I suppose if we put something visual like an icon, it could still go in `<time>` and we just hide it, though then we may want to nest an inline-flex layout or make `<time>` an inline-flex layout.

wdyt @mcarrano @mmenestr, do you ever see putting other stuff in the timestamp component around the timestamp - beside, above, below - like text or a clock icon or actions or something? Or think it would always just be the timestamp?

mcarrano replied:
That's possible. We had some earlier designs with an icon that we abandoned, but it's always possible we'll get a request to extend this.

thatblindgeye commented:
@mattnolting made updates based on your comments:

- Made `.pf-c-timestamp__text` a `<time>` element instead
- Removed the var for cursor property and removed the display property
- Added logic to the template to add a `pf-m-help-text` class and rearranged the scss file to apply the proper styling. Now the underline styling and `cursor: pointer` is only applied when `.pf-c-timestamp` also has the `.pf-m-help-text` class, otherwise only the font size and color are retained.

I kept the previous structure of a timestamp and timestamp__text component for the reasons Michael mentioned in one of the comments above.

mattnolting reviewed (commented):
@thatblindgeye This coming along well. I think `<time class="pf-c-timestamp__text">` is a good addition.

Looks like you want to add `.pf-m-help-text` to the documentation.

thatblindgeye commented on src/patternfly/components/Timestamp/timestamp.scss:
```diff
@@ -0,0 +1,43 @@
+  --pf-c-timestamp__text--Color: var(--pf-global--Color--200);
+
+  // Help text variables for the timestamp text
+  --pf-c-timestamp__text--m-help-text--TextDecorationLine: underline;
```
For the var syntax, I had followed how Progress Stepper and Description List had them setup:

```
--pf-c-description-list__text--m-help-text--cssProperty
--pf-c-progress-stepper__step-title--m-help-text--cssProperty
```

If the `baseComponent--m-help-text__subComponent` syntax would still be better to use, would you want me to add a comment to the progress stepper and description list scss files stating to update them at some point?

mattnolting replied:
Progress stepper and description list help text mod applies to the the text and title elements, whereas `pf-m-help-text` in timestamp applies to the root component rather than `.pf-c-timestamp__text`. I assume the reason for nested elements is that the help text modifier styling applies to all children, including icons etc., otherwise we would use a layout for icon + timestamp.

thatblindgeye replied:
Ah gotcha, that makes sense! Thanks for the clarification, will add that update as well

mattnolting commented on src/patternfly/components/Timestamp/timestamp.scss:
```diff
@@ -0,0 +1,43 @@
+  &.pf-m-help-text {
+    cursor: pointer;
+
+    .pf-c-timestamp__text {
+      text-decoration-line: var(--pf-c-timestamp--m-help-text--TextDecorationLine);
+      text-decoration-style: var(--pf-c-timestamp--m-help-text--TextDecorationStyle);
+      text-decoration-thickness: var(--pf-c-timestamp--m-help-text--TextDecorationThickness);
+      text-underline-offset: var(--pf-c-timestamp--m-help-text--TextUnderlineOffset);
+      text-decoration-color: var(--pf-c-timestamp--m-help-text--TextDecorationColor);
+    }
```
Looks like these styles need to apply to `pf-c-timestamp` rather than `pf-c-timestamp__text`.

mattnolting commented on src/patternfly/components/Timestamp/timestamp.scss:
```diff
@@ -0,0 +1,43 @@
+
+    &:hover {
+      --pf-c-timestamp__text--Color: var(--pf-c-timestamp--m-help-text--hover--Color);
+      --pf-c-timestamp--m-help-text--TextDecorationColor: var(--pf-c-timestamp--m-help-text--hover--TextDecorationColor);
```
applies to `pf-c-timestamp` as well

mattnolting commented on src/patternfly/components/Timestamp/timestamp.scss:
```diff
@@ -0,0 +1,43 @@
+
+    &:focus {
+      --pf-c-timestamp__text--Color: var(--pf-c-timestamp--m-help-text--focus--Color);
+      --pf-c-timestamp--m-help-text--TextDecorationColor: var(--pf-c-timestamp--m-help-text--focus--TextDecorationColor);
```
applies to `pf-c-timestamp` as well

mattnolting commented on src/patternfly/components/Timestamp/timestamp.scss:
```diff
@@ -0,0 +1,43 @@
+.pf-c-timestamp {
+  // Timestamp text
+  --pf-c-timestamp__text--FontSize: var(--pf-global--FontSize--sm);
+  --pf-c-timestamp__text--Color: var(--pf-global--Color--200);
```
I think you can set these props on `pf-c-timestamp` as they're inheritable and any other elements/icons within timestamp would likely need these values as well. Otherwise you would have a dark icon next to lighter text. @mcoker wdyt?

mattnolting reviewed (commented):
My thought is that `pf-c-timestamp__time` would just be a wrapper for the actual `<time>` element and that anything else inside the `pf-c-timestamp` would get the same styling by default.

mcoker reviewed (approved):
LGTM!

mattnolting reviewed (approved):
LPTM! Nice work 👍

patternfly-build commented:
:tada: This PR is included in version 4.205.0 :tada:

The release is available on:
- [GitHub release](https://github.com/patternfly/patternfly/releases/tag/prerelease-v4.205.0)
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/@patternfly/patternfly/v/4.205.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
